client/gutenberg/editor/controller.js

```diff
@@ -291,3 +292,35 @@ export const redirectSiteEditor = async ( context ) => {
 	// Calling replace to avoid adding an entry to the browser history upon redirect.
 	return location.replace( siteEditorUrl );
 };
+/**
+ * Redirect the logged user to the permalink of the post, page, custom post type if the post is published.
+ *
+ * @param {Object} context Shared context in the route.
+ * @param {Function} next  Next registered callback for the route.
+ * @returns undefined      Whatever the next callback returns.
+ */
+export function redirectToPermalinkIfLoggedOut( context, next ) {
+	if ( isUserLoggedIn( context.store.getState() ) ) {
+		return next();
+	}
+	const siteFragment = context.params.site || getSiteFragment( context.path );
+	const CONFIGURABLE_TYPES = [ 'jetpack-portfolio', 'jetpack-testimonial' ];
+
+	// The context.path in this case could be one of the following:
+	// - /page/{site}/{id}
+	// - /post/{site}/{id}
+	// - /edit/jetpack-portfolio/{site}/{id}
+	// - /edit/jetpack-testimonial/{site}/{id}
+	const explodedPath = context.path.split( '/' );
+	if (
+		context.path &&
+		explodedPath[ 1 ] === 'edit' &&
+		! CONFIGURABLE_TYPES.includes( explodedPath[ 2 ] )
+	) {
+		return next();
+	}
+	// Redirect the logged user to the permalink of the post, page, custom post type if the post is published.
+	// else the endpoint will redirect the user to the login page.
+	window.location = `https://public-api.wordpress.com/wpcom/v2/sites/${ siteFragment }/redirect?path=${ context.path }`;
```

Comment:
I just realized that we don't really verify that the site fragment will be properly read as a proper site by the REST API. If this `window.location` redirects to a 404 it would be a pretty bad user experience.

Reply:
A 404 only happends when the site doesn't exits. 
Which should never happen though regular use. A user would have to go and manually change the url.